### PR TITLE
[AD-337] Improve placement of $limit in aggregate pipeline.

### PR DESCRIPTION
**Pull request by birschick-bq**
### Summary

[AD-337] Improve placement of $limit in aggregate pipeline.

### Description

- Add call to estimatedDocumentCount to get more accurate estimate of rows in a table.  https://docs.mongodb.com/v4.0/reference/method/db.collection.estimatedDocumentCount/
- Wrap queries using max row count in `SELECT * FROM () LIMIT` . For queries without limit, this gives an ideal result but for queries that already have a limit clause this still creates duplicates. Extra work still needed here: https://bitquill.atlassian.net/browse/AD-338
- Removed the enumerable rules that are already implemented as DocumentDB rules. 
- Removed the extra $limit for joins by removing the sort join transpose rule. 

### Related Issue

https://bitquill.atlassian.net/browse/AD-337

### Additional Reviewers
@affonsoBQ
@andiem-bq
@alexey-temnikov
@birschick-bq
@mitchell-elholm

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbTable.java [hunk 1]
```diff
@@ -63,13 +65,22 @@
...
     private final DocumentDbSchemaTable tableMetadata;
+    private final Statistic statistic;
 
     protected DocumentDbTable(
             final String collectionName,
             final DocumentDbSchemaTable tableMetadata) {
         super(Object[].class);
         this.collectionName = collectionName;
         this.tableMetadata = tableMetadata;
+        this.statistic = tableMetadata.getEstimatedRecordCount() == -1
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbTable.java]: minor nit.: should we extract -1 as a constant?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 2]
```diff
@@ -221,13 +228,20 @@ public static DocumentDbSchemaTable getTable(
...
                     .collect(Collectors.toMap(
                             DocumentDbSchemaTable::getId,
                             table -> table,
                             (o, d) -> d,
                             LinkedHashMap::new));
+            if (client != null) {
+                for (DocumentDbSchemaTable schemaTable : schemaTables.values()) {
+                    setEstimatedRecordCount(properties, client, schemaTable);
+                }
+            }
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java:243]: What would be scenario when client is null? Should we log a warning?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 3]
```diff
@@ -430,4 +444,22 @@ private static void closeSchemaWriter(final SchemaWriter schemaWriter) throws SQ
...
+                .maxTime(1, TimeUnit.SECONDS);
+        final MongoCollection<Document> collection = client
+                .getDatabase(properties.getDatabase())
+                .getCollection(schemaTable.getCollectionName());
+        try {
+            final long estimatedRecordCount = collection.estimatedDocumentCount(options);
+            schemaTable.setEstimatedRecordCount(estimatedRecordCount);
+        } catch (Exception e) {
+            // NOTE: Not supported before 3.8
+            LOGGER.debug(e.getMessage(), e);
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: Do we want to set default estimated record count in this case?
Have we launched all integration tests on DocumentDB 3.6?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbSchemaTable.java [hunk 4]
```diff
@@ -105,6 +105,11 @@ private static String getSchemaId(final String sqlName, final String uuid) {
     @BsonProperty(COLUMNS_PROPERTY)
     private final List<DocumentDbSchemaColumn> columns;
 
+    @Setter
+    @BsonIgnore
+    @JsonIgnore
+    private long estimatedRecordCount = -1;
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbSchemaTable.java]: Should we use constant (from above?)

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java (hunk 3 above)
Comment (birschick-bq) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: If we can't get an accurate estimate, leave it "unset". Then will follow the path of "UNKNOWN" statistics.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingService.java [hunk 5]
```diff
@@ -114,10 +114,16 @@ public DocumentDbQueryMappingService(final DocumentDbConnectionProperties connec
...
+     * @param maxRowCount the max number of rows to return
      * @return the query context that has the target collection, aggregation stages, and result set metadata.
      */
-    public DocumentDbMqlQueryContext get(final String sql) throws SQLException {
-        final Query<Object> query = Query.of(sql);
+    public DocumentDbMqlQueryContext get(final String sql, final long maxRowCount) throws SQLException {
+        // Add another limit if using setMaxRows.
+        final Query<Object> query =
+                Query.of(
+                        maxRowCount > 0
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingService.java]: nit: I would create a variable assignment for `maxRowCount > 0 ... : sql` - personal style choice.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 6]
```diff
@@ -430,4 +444,22 @@ private static void closeSchemaWriter(final SchemaWriter schemaWriter) throws SQ
...
+            final DocumentDbSchemaTable schemaTable) {
+        final EstimatedDocumentCountOptions options = new EstimatedDocumentCountOptions()
+                .maxTime(1, TimeUnit.SECONDS);
+        final MongoCollection<Document> collection = client
+                .getDatabase(properties.getDatabase())
+                .getCollection(schemaTable.getCollectionName());
+        try {
+            final long estimatedRecordCount = collection.estimatedDocumentCount(options);
+            schemaTable.setEstimatedRecordCount(estimatedRecordCount);
+        } catch (Exception e) {
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: If you could find the more specific exception to catch, that would be better.

Comment (birschick-bq) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: OK. Is compatible with server version 3.6. I think then, it should be OK to no longer catch an exception here and assumen the call to `estimatedDocumentCount(options)` will succeed.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java [hunk 7]
```diff
@@ -348,10 +349,11 @@ private static Operand getMongoAggregateForOrOperator(final RexCall call, final
 
         protected RexToMongoTranslator(final JavaTypeFactory typeFactory,
                 final List<String> inFields,
-                final DocumentDbSchemaTable schemaTable) {
+                final List<String> keys, final DocumentDbSchemaTable schemaTable) {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java:352]: Added keys here so we can access the columns regardless of order. Column map is sometimes changed by $project.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingService.java [hunk 8]
```diff
@@ -114,10 +114,14 @@ public DocumentDbQueryMappingService(final DocumentDbConnectionProperties connec
     /**
      * Uses CalcitePrepare API to parse and validate sql and convert to MQL.
      * @param sql the query in sql
+     * @param maxRowCount the max number of rows to return
      * @return the query context that has the target collection, aggregation stages, and result set metadata.
      */
-    public DocumentDbMqlQueryContext get(final String sql) throws SQLException {
-        final Query<Object> query = Query.of(sql);
+    public DocumentDbMqlQueryContext get(final String sql, final long maxRowCount) throws SQLException {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingService.java:120]: Wrap the query if max rows was set. This is helpful for queries without a limit clause but if there is both limit and max rows there is a duplicate stage (same as current behaviour)

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbTableScan.java [hunk 9]
```diff
@@ -94,6 +95,17 @@ protected DocumentDbTableScan(final RelOptCluster cluster, final RelTraitSet tra
 
         // Keep the project node even for SELECT * queries.
         planner.removeRule(CoreRules.PROJECT_REMOVE);
+
+        // Remove extra $limit on joins.
+        planner.removeRule(CoreRules.SORT_JOIN_TRANSPOSE);
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbTableScan.java:100]: This removes the 1 extra limit on joins. Still need to deal with duplication from max rows + limit.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java (hunk 3 above)
Comment (andiemontoyeah) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: Actually compatible with 3.6. Integration tests are ok on both 3.6 and 4.0

**Review by birschick-bq — COMMENTED** [review]
As I started this PR, I won't be able to approve it.

**Review by alexey-temnikov — APPROVED** [review]

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java (hunk 2 above)
Comment (birschick-bq) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java:243]: It is mainly to deal with testing environment.

**Review by affonsov — APPROVED** [review]

**Pull request merged**